### ISSUE TITLE
Proper way to use search_space

### ISSUE THREAD

**Issue by WDpad753** [hyperopt/hyperopt#818]
I have been working with hyperopt for a while, but, I feel I'm not using the search_space correctly. Here is how I'm using it.

    def search_space(model):
    # Initialising variables:
    model = model.lower()
    space = {}
    # Calling the models:
    if model == 'knn':
        space = {'n_neighbors'      : hp.choice('n_neighbors', [1, 2, 3, 4, 5, 6]),
                 'leaf_size'        : hp.choice('leaf_size', [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99]),
                 'algorithm'        : hp.choice('algorithm', ["auto", "ball_tree", "kd_tree", "brute"]),
                 'metric'           : hp.choice('metric', ["euclidean", "minkowski", "chebyshev", "manhattan"]),
                 'weights'          : hp.choice('weights', ["uniform", "distance"])
                 }

    elif model == 'svc':
        space = {'C'                 : hp.choice('C', [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]),
                 'kernel'            : hp.choice('kernel', ['linear', 'sigmoid', 'poly', 'rbf']),
                 'gamma'             : hp.choice('gamma', [1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
                 }

    elif model == 'dtc':
        space = {'criterion'        : hp.choice('criterion', ['gini','entropy']),
                 'splitter'         : hp.choice('splitter', ['best', 'random']),
                 'max_depth'        : hp.choice('max_depth', [2, 4, 6, 8, 10, 12, 14, None]),
                 'min_samples_split': hp.choice('min_samples_split', [2, 4, 6, 8, 10, 12]),
                 'min_samples_leaf' : hp.choice('min_samples_leaf', [3, 5, 7, 9, 11, 13]),     # If int, then consider min_samples_split as the minimum number.
                 'max_features'     : hp.choice('max_features', ["auto", "sqrt", "log2"]),
                 'random_state'     : hp.choice('random_state', [0])
                 }

    elif model == 'rfc':
        space = {'criterion'        : hp.choice('criterion', ['gini', 'entropy']),
                 'n_estimators'     : hp.choice('n_estimators', [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26]),
                 'max_depth'        : hp.choice('max_depth', [2, 4, 6, 8, 10, 12, 14, None]),
                 'min_samples_split': hp.choice('min_samples_split', [2, 4, 6, 8, 10, 12]),
                 'min_samples_leaf' : hp.choice('min_samples_leaf', [3, 5, 7, 9, 11, 13]),     # If int, then consider min_samples_split as the minimum number.
                 'max_features'     : hp.choice('max_features', ["auto", "sqrt", "log2"]),
                 'random_state'     : hp.choice('random_state', [0])
                 }

    elif model == 'etc':
        space = {'criterion'        : hp.choice('criterion', ['gini', 'entropy']),
                 'n_estimators'     : hp.choice('n_estimators', [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26]),
                 'max_depth'        : hp.choice('max_depth', [2, 4, 6, 8, 10, 12, 14, None]),
                 'min_samples_split': hp.choice('min_samples_split', [2, 4, 6, 8, 10, 12]),
                 'min_samples_leaf' : hp.choice('min_samples_leaf', [3, 5, 7, 9, 11, 13]),     # If int, then consider min_samples_split as the minimum number.
                 'max_features'     : hp.choice('max_features', ["auto", "sqrt", "log2"]),
                 'random_state'     : hp.choice('random_state', [0])
                }

    elif model == 'xgbc':
        space = {'booster'          : hp.choice('booster', ['gbtree', 'gblinear']),
                 'objective'        : hp.choice('objective', ['reg:linear', 'binary:logistic', 'multi:softmax', 'multi:softprob']),
                 'max_depth'        : hp.choice('max_depth', [2, 4, 6, 8, 10, 12, 14, None]),
                 'learning_rate'    : hp.choice('learning_rate', [0.05, 0.10, 0.15, 0.20, 0.25, 0.30]),
                 'min_child_weight' : hp.choice('min_child_weight', [1, 3, 5, 7]),
                 'gamma'            : hp.choice('gamma', [0.0, 0.1, 0.2 , 0.3, 0.4]),
                 'colsample_bytree' : hp.choice('colsample_bytree', [0.3, 0.4, 0.5 , 0.7])
                }

    elif model == 'adabc':
        space = {'n_estimators'     : hp.choice('n_estimators', [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48]),
                 'learning_rate'    : hp.choice('learning_rate', [0.05, 0.10, 0.15, 0.20, 0.25, 0.30]),
                 'algorithm'        : hp.choice('algorithm', ["SAMME", "SAMME.R"]),
                 'random_state'     : hp.choice('random_state', [0])
                }

    space['model'] = model
    return space

This function works okay, but, My following questions:

- for range parts like _hp.choice('n_estimators', [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48])_, is it best to use quniform?
- In search space, is it possible to use both None and like range of values together? For example, _'max_depth'        : hp.choice('max_depth', [2, 4, 6, 8, 10, 12, 14, None]),_
- _Side question_, if I want to use cross-validation, is the only place to use it in the objective function?

**Comment by imbalu007** [comment]
It would be nice to have a hyperopt API that validates the input space. It would help us make sure our space definition is correct

**Comment by WDpad753** [comment]
>  In search space, is it possible to use both None and like range of values together? For example, 'max_depth' : hp.choice('max_depth', [2, 4, 6, 8, 10, 12, 14, None])

I managed to sort this one by doing the following:

`
# Create an empty list:
max_depth_vl = []
lst_1 = np.arange(2, 16, 1).tolist()
lst_2 = [None]
max_depth_vl.extend(lst_1)
max_depth_vl.extend(lst_2)

# input list into space:
...... area where max_depth => hp.choice('max_depth', max_depth_vl)
`
and

> for range parts like hp.choice('n_estimators', [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48]), is it best to use quniform?

I thought of using `np.arange(2, 16, 1).tolist()` in places with long list of values, however, I am still wondering if quniform is still best to use?

**Comment by michaelmior** [comment]
I don't believe you'll see any significantly different behavior using `quniform` and it certainly makes the code easier to read than using a long list of values. Keep in mind though that `quniform` produces floating point values, so you'll have to convert to integers if that's what you want.
